Modularize javascript

HelgeKrueger opened the pull request:
This pull request splits the main.js file into several node modules and uses browserify to recreate main.js. With node.js installed main.js can be created using `npm run bundle` inside the `web` folder.

See discussion at https://discuss.graphhopper.com/t/modularize-javascript-code/306/16

Installation:

```bash
curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.29.0/install.sh | bash
# close and reopen terminal now
nvm install 4.2.2
nvm use 4.2.2

# git clone https://github.com/graphhopper/graphhopper.git
cd graphhopper/web
# download required packages:
npm install
npm test
# create main.js
npm run bundle

cd ..
./graphhopper.sh web africa_canary-islands.pbf
firefox localhost:8989
```

If you wanted to add the underscore library to the project, then you would use `npm install --save underscore` command. This command automatically modifies package.json to include underscore and a version info. This means that the next person running nvm install would also download underscore.

```bash
# For development just use watchify:
npm run watch
# see the package.json where several other scripts like bundleDebug are defined
# bundle creates the main file
npm run bundle
# create main.js for debugging
npm run bundleDebug
npm run bundleUgly

# Forcing consistent code style with jshint. JsHint can be run with:
npm run jshint
```

karussell commented:
Thanks again :) ! 

Do we still provide an out of the box compiled main.js? Or do you think that this is a good solution?

HelgeKrueger commented:
My last commit looks odd. It was meant to include the precompiled minified main.js. Let me try to fix it.

HelgeKrueger commented:
I added the jasmine test to travis-ci and added jshint, which is style checker for JavaScript. I also added various missing semicolons, that jshint found.

karussell commented on .travis.yml:
```diff
@@ -5,10 +5,12 @@ jdk:
-script: ./core/files/travis-build.sh
+script:
+  - ./core/files/travis-build.sh
+  - ./web/files/travis-build.sh
```
Hmmh, as the web jar is also built in the `core/files` we should either avoid a separate `web/files` script or put the web jar building into the `web/files` I fear :)

karussell commented on web/package.json:
```diff
@@ -0,0 +1,27 @@
+    "lint": "jshint src/main/webapp/js/"
+  },
+  "author": "",
+  "license": "ISC",
```
We should use Apache License 2.0 here as for all other GH stuff.

karussell commented:
Thanks again for this work and also for making the tests working even on travis :) ! Is npm installed per default there?

Is leaflet.loading.js removed? See the functionality: https://github.com/ebrelsford/Leaflet.loading

HelgeKrueger commented:
npm is installed by default on travis-ci.

HelgeKrueger commented on web/package.json:
```diff
@@ -0,0 +1,27 @@
+    "jasmine": "^2.3.2",
+    "jquery": "^2.1.4",
+    "leaflet": "^0.7.7",
+    "leaflet-loading": "^0.1.15",
```
Leaflet loading inclusion part 1.

HelgeKrueger commented on web/src/main/webapp/js/main-template.js:
```diff
@@ -0,0 +1,591 @@
+
+global.d3 = require('d3');
+var L = require('leaflet');
+require('leaflet-loading');
```
leaflet-loading inclusion part 2.

I was not able to use this npm inclusion mechanism for the other leaflet / jquery plugins. I also don't feel that this is super important at the moment to get all packages via npm.

karussell commented:
Thanks!

I'm still struggeling to understand how I would enable 'debug mode'. It is still possible to change something in the source code and hit F5 without a build command in-between? 

And also I've added a bit installation instructions in your initial comment of this PR ... would you mind to explain the different bundle commands (see TODO)?

karussell commented:
It looks like watchify is the solution (?)

```bash
npm install -g watchify
watchify src/main/webapp/js/main-template.js -o src/main/webapp/js/main.js
```

Is this good or what about beefy?

karussell commented:
Merged it here https://github.com/graphhopper/graphhopper/commit/38e1c32069d1f296bbcbf3b5a09f00a6389b764c

One open question for me: "Also should we rename main.js to bundle.js and main-template.js to index.js or main.js to be more align with the normal namings for npm projects?"
